### PR TITLE
🌱  Enable godoc linter checks

### PR DESCRIPTION
**Pull request by stmcginnis**
**What this PR does / why we need it**:

This is a continuation of the work being done to enable some of the excluded linters we use.

This change removes the exclusions related to enforcing godoc comments for packages and
functions, variables, and other things that are public. It makes they all have associated doc
strings and that any package comments are in the expected format of `Package x...`.

For package docs, if a given package was less than five separate files the comments were just
added to the `package` statement in the first file. If there were five or more, a `doc.go` file was
added to keep it separate and easy to find.

**Which issue(s) this PR fixes**:
Related #4622

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -48,8 +48,6 @@ issues:
     - Error return value of .((os\.)?std(out|err)\..*|.*Close|.*Flush|os\.Remove(All)?|.*print(f|ln)?|os\.(Un)?Setenv). is not checked
     # The following are being worked on to remove their exclusion. This list should be reduced or go away all together over time.
     # If it is decided they will not be addressed they should be moved above this comment.
-    - (comment on exported (method|function|type|const)|should have( a package)? comment|comment should be of the form)
-    - package comment should be of the form "Package (.+) ..."
```
Comment (vincepri) [inline .golangci.yml:52]: Let's keep this exclusion? Seems a lot of comment are in the form of `package X implements X`, which isn't really super useful?

File: hack/boilerplate/test/fail.go [hunk 2]
```diff
@@ -1,8 +1,6 @@
 /*
 Copyright 2014 The Kubernetes Authors.
 
-fail
```
Comment (fabriziopandini) [inline hack/boilerplate/test/fail.go:4]: I think this is required to get the boilerplate tests to fail

File: .golangci.yml (hunk 1 above)
Comment (fabriziopandini) [reply .golangci.yml:52]: I don't have strong opinions here.
Ideally we should get few lines explaining what each package is, but the linter just ensures a minimal comment exist, not that an actual explanation exists.
However, if we consider this as a starting point I'm ok with enabling this linter/this PR

Comment (vincepri) [reply .golangci.yml:52]: My worry is that we add these descriptions to satisfy the linter, and provide no information to our users.

Comment (stmcginnis) [reply .golangci.yml:52]: We could keep the regex for just packages, but I would actually like to keep this.

I agree most of the docstrings are not very helpful right now. But at least they have a basis to be improved upon. There are a couple cases where it actually did add some helpful information and caught where there were good godocs, but not formatted correctly to show up in IDE help because of where they were placed. And it would help make sure any new packages added at least force the authors to think about whether there is anything helpful to add to the godocs.

So bottom line, the added value is questionable, but it at least has potential for helping without much of a downside. Does that make sense to you @vincepri?

File: hack/boilerplate/test/fail.go (hunk 2 above)
Comment (stmcginnis) [reply hack/boilerplate/test/fail.go:4]: Ah, that explains it. I could tell why that was there and assumed it was a copy/paste error.

I will wait to get Vince's feedback on other changes, then get this updated.

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Comment by vincepri** [comment]
Need a rebase

**Comment by stmcginnis** [comment]
/test pull-cluster-api-test-main

Unrelated failures.

**Comment by stmcginnis** [comment]
Merge conflicts resolved and tests are happy again. Please take a look again when you have some time @vincepri - thanks!
